Contact Cards

vitorpamplona opened the pull request:
Uses unbound lists to create events that define the relationship between two pubkeys in a public or private manner.

Read [here](https://github.com/vitorpamplona/nips/blob/relationship-status/81.md)

vitorpamplona committed:
First Draft

vitorpamplona committed:
Quick adjustments

vitorpamplona committed:
adds privately

sudocarlos committed:
correct typo

jb55 commented:
cc @Semisol

alexgleason commented:
Will this allow for `Alex Gleason set their relationship status to "it's complicated" with X`?

arthurfranca commented:
Previous discussion regarding contact list at #349
tldr: see "kind X follow (by @vitorpamplona)" section [here](https://github.com/nostr-protocol/nips/pull/349#issuecomment-1472395508)

vitorpamplona committed:
Merge pull request #1 from sudocarlos/patch-1

correct typo

fiatjaf commented:
@alexgleason I think [NIP-32](https://github.com/nostr-protocol/nips/blob/master/32.md) may be the best way to declare public relationship statuses.

vitorpamplona commented:
There are so many jokes here that I don't know if this proposal has any legs 😅

arthurfranca commented:
Most "Relationship Status" lists will have few items so NIP-51 would be enough.
Hopefully I expect my unbounded list PR #784 to be at a good enough state to suggest holding some ever growing data like subscriber list.

@vitorpamplona Please take a look at this example I just added there:

https://github.com/nostr-protocol/nips/blob/8ae2b5d61b47c3fe9ca6781dd76ced1693a550e6/61.md?plain=1#L53-L87

vitorpamplona commented:
So, the goal for Relationship Status is that most of us have over 10,000 contacts (not friends, not follows, just contacts). And a list with 10,000 people is quite a heavy list to be updated all the time. That was the reason to start this.

Like, an email client, for instance, can have a status group for every person that sent you an e-mail.

arthurfranca commented:
So, with my unbounded list proposal (basically the `n` tag and the 33118 kind for "people") it could be like this:

```js
{
  "kind": 33118,
  "tags": [
    ["d", "random6467362966154779"]
    ["n", "email_contacts"]
  ],
  "content": "<NIP-44 encrypted Stringified TAG-List([
    ["p", "91cf9..4e5ca", "wss://alicerelay.com/", "<nickname>", "<summary>"]
  ])>",
  ...other fields
}
```

The difference is that you can later add the same entry simultaneously to another unbounded list editing it and adding a new `n` tag.

Want all email contacts, search by the n tag and author.

arthurfranca commented:
@vitorpamplona Added you as author and also the "contacts" example there at #784
Please check it out.

vitorpamplona commented:
> The difference is that you can later add the same entry simultaneously to another unbounded list editing it and adding a new n tag.

I am confused. What does it do? Why would I add the entry to another unbounded list?

arthurfranca commented:
> you can have multiple p-tags in kind: 33118

A kind 33118 event was meant to have a single `p` tag (I have recently edited the description pointing that out), because an unbounded people list (with a specific `n` tag) is made of one to many of these events.
There is another example there with two `p` tags but it is the NIP-51 kind 30000 with an `n` tag.

> You could turn the status into the unbound event's n tag... to get a single person's status by filtering by kind:30382, p tag and pubkey.

Hmm got it you want to have kind 30382 with a p tag unencrypted (to able to use on filter). I encrypted the `n` tag content
to hide the info there.

```js
{
  "kind": 30382, // can fetch all contacts
  "tags": [
    ["d", "random5189992173907396"]
    ["n", <NIP-44 encrypted("known_face")>], // can fetch all "known_face"s but no one knows the status
    ["p", "91cf9..4e5ca", "wss://alicerelay.com/"],
  ],
  "content": "<NIP-44 encrypted Stringified TAG-List([
    ["nickname", "<My buddy>"],
    ["summary", "<summary of the relationship>"],
    ["email", "bla@ble.com"]
  ])>",
  ...other fields
}
```

Is this it?

arthurfranca commented:
If the above `["n", <NIP-44 encrypted("known_face")>]` is a good idea, maybe you should standardize a NIP-51 list
holding all used `n` statuses?

I encrypted because someone may get mad if it discovers his friend labeled him as a "scammer" or something worse =o

vitorpamplona commented:
ohh interesting.. i didn't know the `33118` was supposed to be for just one person. I thought it would be a list of lists of people. Which is probably useful in some use cases. 

`["n", <NIP-44 encrypted("known_face")>]`

Maybe using status for `n` is not a good idea. I am not sure if the encryption is deterministic to make sure all `known_face`s map to the same base64.

arthurfranca commented:
> Maybe using status for n is not a good idea. I am not sure if the encryption is deterministic to make sure all known_faces map to the same base64.

Just tested it. It is not deterministic =(

```
nip44.encrypt(hexToBytes(sk1), 'scammer') // 'AU9OCx1Rn7PAA5TxDqgi5Up7bK+cNAgI2bU5/BhQmg0='
nip44.encrypt(hexToBytes(sk1), 'scammer') // 'AfCaDG/6jzFvQKEAVkaKzlgKIsPddNWXaynDPPNb/64='
```

arthurfranca commented:
> maybe you should standardize a NIP-51 list holding all used n statuses?

If searching by the status at `n` tag was a must have feature, it could still be done if such NIP-51 list (holding all used n statuses) had tags like this:

```js
[
  ["status", "<status 1>", "random string 1"],
  ["status", "Big Time Scammer", "6064460175057025"]
]
```

Then the `n` tag would be "6064460175057025"

vitorpamplona committed:
Refines proposal to make relationships private.

vitorpamplona committed:
Improves formatting

vitorpamplona committed:
Optional n tag

vitorpamplona committed:
correcting the example

vitorpamplona commented:
@arthurfranca I am back to this, changed it to a private status (hash of the key as `d`-tag) and changed it to your unbound list idea. 

I will be using this to document a list of Patients a doctor is taking care of. People can count how many patients a doctor has, but the relationship between them should be fully private.

vitorpamplona committed:
Switch to random UUID for better security.

arthurfranca commented on 81.md:
```diff
@@ -0,0 +1,38 @@
+NIP-81
+======
+
+Private Relationship Status
+---------------------------
```
What about renaming it to "Private Contacts"?

arthurfranca commented on 81.md:
```diff
@@ -0,0 +1,38 @@
+  "kind": 30382,
+  "tags": [
+    ["d", "<randomUUID>"],
+    ["n", "Clients"]
```
What do you think of defining a NIP-51 list kind to hold all possible statuses? It would be easier to help user pick
a status label when adding/editing a contact. It would be possible to set the `n` tag to a private status ("6064460175057025" instead of "Idiot").

```js
{
  "kind": 1XXXX,
  "content": "<nip44Encrypt(JSON.stringify([
    ["status", "<status 1>", "random string 1"],
    ["status", "Clients", "Clients"] // public
    ["status", "Idiot", "6064460175057025"] // private
  ]))>",
  // ...other fields
}
```

vitorpamplona replied:
I think `n` itself can be public or private. Clients can always show the `n` value if that value is not found in the kind 10XXXX list

vitorpamplona replied:
ohh interesting.. I will add it to the 10xxx events as well

vitorpamplona commented on 81.md:
```diff
@@ -0,0 +1,38 @@
+{
+  "kind": 30382,
+  "tags": [
+    ["d", "<randomUUID>"],
```
I was trying to hash as well but I wasn't completely sure it was safe given that a similar process is used to determine the private conversation key between two pubkeys using NIP44

arthurfranca replied:
I think hmac with sha256 and nostr privkeys have the same number of possible combinations so it would be just as hard to crack it, so using hmac wouldn't be lowering the security of the privkey. Though I may be wrong @paulmillr @mikedilger @Semisol to the rescue?

paulmillr replied:
yeah this is a bad idea. 1. hkdf should be used for these cases instead of hmac. 2. keys should not be reused between protocols (MAC vs elliptic curve key).

arthurfranca replied:
@paulmillr got confused with the 2nd statement. Like can we safely use hkdf with nostr privkey to deterministically hash a text? If yes, can you please give us an example of how would it be with your '@noble/hashes/hkdf' lib?

paulmillr replied:
why does it need to contain user's private key and not public?

vitorpamplona replied:
> why does it need to contain user's private key and not public?

Too easy to find out the hash of two pubkeys used in Nostr. You can just have a big table of hashes for every 2 USED pubkeys and know who is talking to who (which should be private).

arthurfranca replied:
The main idea is to hide from public a string value in a way no one without something only I have (my privkey) can guess the original value by hashing a list of strings the person thinks may include the original string.

I mean, if I have a list of people I follow (pubkeys), if I hash a pubkey simply with `sha256(string)`, people will easily guess which one by doing sha256 on all my follow list.

paulmillr replied:
I suggest this:

```
nip81_key = hkdf(private_key, salt: 'nip81') 
bobs_tag = sha256(nip81_key || bobs_pub)
kates_tag = sha256(nip81_key || kates_pub)
```

paulmillr replied:
the less stuff we do with private keys, the less chance it would accidentally leak, etc. nip81 key can be cached and passed anywhere, since it's less important

vitorpamplona committed:
Updating NIP-81 to add private and public statuses

vitorpamplona committed:
formatting

vitorpamplona committed:
enhancing the example with private ntags

vitorpamplona committed:
wording

vitorpamplona committed:
formatting

vitorpamplona committed:
formatting

vitorpamplona committed:
Merge remote-tracking branch 'upstream/master' into relationship-status

vitorpamplona committed:
Moves to 10008 kind for name maps

vitorpamplona committed:
Adds edge cases

vitorpamplona commented on 81.md:
```diff
@@ -0,0 +1,90 @@
+
+The private tags are JSON Stringified, [NIP-44](44.md)-encrypted and placed inside the `.content` of the event. 
+
+Example of Public Status
```
Both should be available. Anyone building web of trust needs these graph edges very well defined.

fr4nzap replied:
Can you expand on `trust_level`? Is it optional? Is it a range between 0 and 1? Why is it a string?

vitorpamplona replied:
It is optional and undefined as of now (it was there just to give an idea that it was possible). I thought of it as a normalized score/weight from one of these algorithms.

vitorpamplona committed:
renames nicknames to petnames

vitorpamplona committed:
Simplifying statuses

vitorpamplona committed:
Creating two event kinds.

arthurfranca commented on 81.md:
```diff
@@ -0,0 +1,73 @@
+  "kind": 30382,
+  "tags": [
+    ["d", "e88a691e98d9987c964521dff60025f60700378a4879180dcbbb4a5027850411"],
+    ["n", "Follows"],
+    ["n", "Bitcoiners"],
```
```suggestion
    ["n", "follow"],
    ["n", "bitcoiner"],
```

arthurfranca replied:
Can we convention unbound list names to using lowercase, singular form, English words separated by dash?
Added this info to the the unbound list NIP.

arthurfranca commented on 81.md:
```diff
@@ -0,0 +1,73 @@
+  "kind": 10008,
+  "content": nip44Encrypt(JSON.stringify([
+    ["map", "<code>", "<name>"],
+    ["map", "Bitcoiners", "Bitcoiners"], 
+    ["map", "6064460175057025", "Debtors"]
```
```suggestion
    ["map", "bitcoiner", "bitcoiner"], 
    ["map", "6064460175057025", "debtor"]
```

arthurfranca commented on 81.md:
```diff
@@ -0,0 +1,73 @@
+  ],
+  "content": nip44Encrypt(JSON.stringify([
+    ["p", "<pubkey>", "relay"],
+    ["n", "Clients List"],
```
```suggestion
    ["n", "clients-list"],
```

vitorpamplona committed:
Update 81.md

Co-authored-by: arthurfranca <arthur.a.franca@gmail.com>

vitorpamplona committed:
Update 81.md

Co-authored-by: arthurfranca <arthur.a.franca@gmail.com>

vitorpamplona committed:
Update 81.md

Co-authored-by: arthurfranca <arthur.a.franca@gmail.com>

vitorpamplona committed:
Merge remote-tracking branch 'upstream/master' into relationship-status

# Conflicts:
#	README.md

vitorpamplona committed:
Updates relationship status to use event sets.

vitorpamplona committed:
adds public relationship status to the readme

vicariousdrama commented:
I'm in favor of this NIP.

Corny Chat implemented basic support for NIP-81, kind 30382 on 2024-05-06 for petnames. This is an exceptionally useful feature so that users can actually start giving personal names to pubkeys which helps resolve who's who amongst the Dans, and acts as a defence against impersonation.  NIP-44 encryption is used for those whose extensions support it. Otherwise it falls back to plain text.

vitorpamplona commented:
@alexgleason did you add 30382 based on this NIP? 

@vicariousdrama @alexgleason can you check if the NIP text matches your implementations? If not, I can change the text.

alexgleason commented:
@vitorpamplona You can run `nak req -k 30382 -a db0e60d10b9555a39050c258d460c5c461f6d18f467aa9f62de1a728b8a891a4 wss://gleasonator.dev/relay` to see my events.

I am using custom `n` values, none of which are described here. Example:

```json
{
  "kind": 30382,
  "id": "3a7f95cb6b72a2ebdac6c3f1410e33685efb2644ae18e7be1f5f8933a8f4d59b",
  "pubkey": "db0e60d10b9555a39050c258d460c5c461f6d18f467aa9f62de1a728b8a891a4",
  "created_at": 1718215730,
  "tags": [
    ["d", "932614571afcbad4d17a191ee281e39eebbb41b93fac8fd87829622aeb112f4d"],
    ["n", "admin"],
    ["n", "suggest"],
    ["n", "suggested"]
  ],
  "content": "",
  "sig": "ca77c7313785f39434c1b944865fcce73f1053947e69ef6667e7c76a2ef8af70af6d1d37cd6cc8e530907868ffeec85412d0e6e9c7f973b29a5de6b2cc3be2b7"
}
```

- `admin` - the user is granted admin permission of the ditto server, since this event is signed by the [Ditto nsec](https://docs.soapbox.pub/ditto/nsec)
- `suggested` - the user is globally suggested in the UI
- `suggest` (deprecated) - I decided to go with `suggested` (even though it doesn't match the tense of `follow`) and forgot to remove the tag from existing events.

I am using this in a slightly unconventional way, but it all makes sense. I intend for regular users to also create kind 30382 events for things like private notes about users, and for users to suggest other users in future updates.

vitorpamplona commented:
Nice! 

I am not sure if the `n`s should be defined here. I am just making sure the rest of it is like we described. :)

vitorpamplona committed:
Aligning the Private Relationship status Kind with the other PRs

vitorpamplona committed:
Fixing kind numbers

arthurfranca commented:
> Fixing kind numbers

:+1:

vitorpamplona commented on README.md:
```diff
@@ -172,11 +172,11 @@ They exist to document what may be implemented by [Nostr](https://github.com/nos
 | `30403`       | Draft Classified Listing   | [99](99.md)              |
 | `30617`       | Repository announcements   | [34](34.md)              |
 | `30818`       | Wiki article               | [54](54.md)              |
+| `31382`       | Private Relationship Status| [81](81.md)              |
```
No that's for the event, according to the old version of #784

mikedilger replied:
but there is no reference to 31383 in this PR, so then it shouldn't be in the README either unless it is explained.

vitorpamplona replied:
I am confused, 31383 is not in the readme. 

Public  30382
Private 31382

vicariousdrama commented:
> @alexgleason did you add 30382 based on this NIP?
> 
> @vicariousdrama @alexgleason can you check if the NIP text matches your implementations? If not, I can change the text.

I've checked how I've implemented this in Corny Chat, and only use kind 30382.  The `petname` tag is used, as a tag, or in encrypted content via nip44 if the client signer supports it. This appears to match how your nip is written

Giszmo commented:
ACK. Looks good to me.

SilberWitch commented:
I wish I could link different sets of profile data with different NIP-05s, so that the display changes by community, with a bland/anon one for Twitter clones.
Could we use kind 30382 for that?

SilberWitch commented:
Is there a reason why the `d` tag is used to reference an npub, in kind 30382? Shouldn't that be a `p` tag, since it's a reference to another user, and we generally use `d` tags as an identifier for long-form notes?

vitorpamplona commented:
> Is there a reason why the d tag is used to reference an npub, in kind 30382?

It's so that you can only have one 30382 per key.

SilberWitch commented:
Ah. I see. 🤔
I thought it was a typo.

That makes sense, since they're replaceable events, but a bit confusing.

vitorpamplona committed:
Merge remote-tracking branch 'upstream/master' into relationship-status

# Conflicts:
#	README.md

vitorpamplona committed:
moves event set names to the right position on readme

companion256 commented:
What's the status of this PR? Seems stale

vitorpamplona commented:
> What's the status of this PR? Seems stale

It's is done as is. Ditto and a few other clients implement this. Amethyst is on draft yet.

SilberWitch commented:
Then why isn't it merged, if it's already in active use by multiple clients? Have to dig through the PRs, to find it.

vitorpamplona commented:
> Then why isn't it merged, if it's already in active use by multiple clients?

Merging things too early in a core proposal (it could replace contact lists) like this one is bad. 

I am ok with waiting for more people to code and be convinced that this is a good approach or not.

vitorpamplona commented:
> Just means that protocol changes in the repo are too difficult and slow

Protocol changes should be slow. New usage of the protocol can be much faster. Most NIPs are just new usages. 

> the development of the protocol always happens elsewhere

As it should. The protocol should be developed, tested and show some proven usage before the specs are merged. Otherwise, we would have had 1000s of NIPs already and 90% would be unused. 

> repo is now just an archive of old versions.

This repo does not direct where Nostr is going. It only reflects what people are actually doing and using. PRs are just propositions of where apps want to go to allow other devs to react to any blatant issue when encoding things into Nostr.

SilberWitch commented:
Okay, but then, what's the advantage of keeping the spec in the repo, versus just having a NUD in the wiki and listing the corresponding event kind in the repo ReadMe? Everything in here is outdated.

vitorpamplona commented:
> what's the advantage of keeping the spec in the repo, versus just having a NUD in the wiki and listing the corresponding event kind in the repo ReadMe

There is no significant advantage for the NIP. It's just additional exposure. In fact, many people would say NUDs are a better system for specs than a centralized repo run by Microsoft. 

> Everything in here is outdated.

Outdated means obsolete. Which is not the case. This repo just doesn't have all the crap people are testing, kick-starting, etc. It only has the things that have shown they are useful. 

Think about this repo as an interoperability layer. If multiple clients are trying to be interoperable, this is a good spot to come together. If a client is not interested in getting others to use their stuff, they don't need to be here at all. In fact, many couldn't care less about interoperability.

SilberWitch commented:
We low-key hate GitHub, but Nostr development has kind of forced us to use it. We were usually using other git servers, before.

jb55 reviewed (approved):
its a good idea

staab commented:
I would use `l`

vitorpamplona commented:
We had that discussion before. Just because they have the same format doesn't mean they are the same. 

1. I don't think we need `kind:10008`s name overrides for NIP-32 labels. Mixing the two tags by declaring here that we are using NIP-32 labels could just confuse everyone that kind 10008's overrides apply to all NIP-32 labels, which is not the case. 
 
2. NIP-32 Labels are designed to have some notion of meaning that is larger than a pubkey. That's why it focuses on ontologies so much. Each user may have his/her own interpretation of the ontology but it is beneficial to NIP-32 if labels remain constrained to the classification/ontology work, even on optional name spaces. `n` tags are completely unbound to anything. Only their public key knows their true meaning. Users can and will lie on those names on purpose.

The semantical difference between a hashtag, a label, and a set is big enough to justify their separate existences.

vitorpamplona commented:
> global tags

There is no such thing as a "global tag" on Nostr. Each NIP redefines every tag, both syntactically and semantically, to their own liking. Most of the redefinitions are syntatically backwards-compatible but semantically quite different. In code, you can never expect a global tag approach to last.

> merged all the concepts

There is no point in merging concepts. What's to gain from it? You lose all the beautiful specialization that each of them have for the sake of what? One less character to use on your source code? One less if (if "t" -> if "t" or "l")? 

Generalizations are the enemy of interoperability.

So, no I don't think we should merge `t` and `l` until we find some massive gains for the user.

vitorpamplona committed:
Merge branch 'master' into relationship-status

vitorpamplona committed:
Changes to Contact Cards

vitorpamplona committed:
Update 81.md
